processing/p5.js-web-editor
# Adds Personal Access Tokens for API access (fixes #541)

Implements Personal Access Tokens for API access. This builds on top of the excellent work in #731 to allow access tokens to be created by users and used to authenticate instead of a password.

I plan on making the following changes from #731:

 - [x] Move token generation to the server-side from the client. This seems to be how other providers do this (Digital Ocean, Github, Harvest).
- [x] Displays tokens as part of Account page (moving from the Advanced page)
- [x] Write UserController tests for creating/deleting keys
- [x] Improve display of tokens in UI
- [ ] Write UI tests **<- I'll do these in a future PR**

I have verified that this pull request:
* [x] is from a uniquely-named feature branch and has been rebased on top of the latest master. (If I was asked to make more changes, I have made sure to rebase onto master then too)

## UI

I've added tabs to the Account screen. The _Access Tokens_ tab allow tokens to be created and deleted.

<img width="888" alt="User interface for managing Access Tokens" src="https://user-images.githubusercontent.com/1762/57783368-4c129d00-772e-11e9-8528-57a5baead215.png">

## Automated testing

I've written tests for the controller endpoints but created Mongoose model mocks by hand. This isn't ideal so I'd like to re-visit these in a future PR once we've figured out a good way forward for testing.

## Manual test plan

1. Create a new access token at /account
2. Copy access token and use `curl` to access the test endpoint:
    `curl -X GET --user <username>:<access-token> -i localhost:8000/api/auth/access-check`
    If successful, the user's session data should be returned and _Last used_ time in UI should update.
3. Change username, access should be forbidden
4. Change token, access should be forbidden
5. Delete token via UI
6. API access should be forbidden
